olive/common/utils.py

```diff
@@ -415,3 +415,28 @@ def find_first_matched_value(original_dict: Dict, keys: Union[str, Tuple, List[s
     if raise_key_error:
         raise KeyError(f"Keys {keys} not found in {original_dict}")
     return None
+
+
+def get_credentials(default_auth_params: Dict = None):
+    """Get credentials for MLClient.
+
+    Order of credential providers:
+    1. Azure CLI
+    2. DefaultAzureCredential
+    3. InteractiveBrowserCredential
+    """
+    from azure.identity import DefaultAzureCredential, InteractiveBrowserCredential
+
+    logger.debug("Getting credentials for MLClient")
+    try:
+        default_auth_params = default_auth_params or {}
+        credential = DefaultAzureCredential(**default_auth_params)
+        # Check if given credential can get token successfully.
+        credential.get_token("https://management.azure.com/.default")
+        logger.debug("Using DefaultAzureCredential")
+    except Exception:
+        logger.warning("Using InteractiveBrowserCredential since of default credential errors", exc_info=True)
+        # Fall back to InteractiveBrowserCredential in case DefaultAzureCredential not work
+        credential = InteractiveBrowserCredential()
```

Comment:
`InteractiveBrowserCredential` is only valid if launched in interactive mode. On CI this will result in infinite wait. This should be wrapped in a condition of whether the launch is interactive or not. In other cases, this should just fail.

Reply:
For automation purposes, we could consider using some environment variable that is available only on CI. This would be default and can be overriden by a command line argument.

Reply:
How does current pipeline work with `InteractiveBrowserCredential()`? I assume it will timeout and fall back to next credential.

Reply:
Apparently, as per the documentation [here](https://learn.microsoft.com/en-us/dotnet/api/azure.identity.interactivebrowsercredential?view=azure-dotnet) there is no timeout or anyway to cancel it unless the request is launched as an async job. I guess, it's expected as the name says _interactive_. If there are other places using the same/similar calls, they are all wrong. Those calls will wait indefinitely for user action.

Reply:
Please add a task in our backlog to revisit the use of `InteractiveBrowserCredentials`. I don't think its use in the codebase is correct. Resolve the conversation with link to the task.

Reply:
Done https://aiinfra.visualstudio.com/Model%20optimization%20Toolkit/_workitems/edit/38586